### PR TITLE
multi-dim standard deviation for CUDA.

### PR DESCRIPTION
**Pull request by umanwizard**
This is the CUDA version of #14535 .
It refactors Reduce.cuh to allow more general classes of reductions to be performed -- we no longer assume that the temporary data returned during reduction is just one scalar, and instead allow an arbitrary accumulate type.
We also allow 64-bit indexing when necessary, since in general we will no longer be able to accumulate directly in the output. (In the cases when we can, we continue to split the tensors until they can be addressed with 32-bits, as before).
As an initial use-case, we implement `std` in multiple dimensions.

### CODE REVIEW

**Comment by umanwizard** [comment]
`var` based on this: (Will open a PR against the main PyTorch repo once this one lands) https://github.com/umanwizard/pytorch/pull/1

**Comment by ezyang** [comment]
cc @iotamudelta @bddppq about ROCm breakage.

**Comment by ezyang** [comment]
I figured it out: the problem is `aten/src/ATen/native/Welford.h` is not getting HIPified.

**Comment by ezyang** [comment]
Suggested resolution is to make the includes point in the right place (THH.h for ROCM, rather than THC.h, e.g.)

**Comment by umanwizard** [comment]
Timings: https://gist.github.com/umanwizard/80e23780ef021ed9a07cb454f2cd4956

TLDR: `sum` hasn't changed. `std` is now always at least as good, and sometimes multiple times faster.

**Review by colesbury — APPROVED** [review]
